File: CppSolution/BinaryTreeLevelOrderTraversal2.cpp
#include <iostream>
#include "Utility.h"
#include <queue>
using namespace std;

class Solution 
{
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) 
    {
    	queue<TreeNode*> tq;
    	queue<int> lq;
    	vector<vector<int> > res;
    	vector<int> tmp;
    	if (root == NULL)
    		return res;
    	tq.push(root);
    	lq.push(1);
    	int level;
    	int lastlevel = 1;
    	TreeNode *p;
    	while (!tq.empty())
    	{
    		level = lq.front();
    		lq.pop();
    		if (level > lastlevel)
    		{
    			res.push_back(tmp);
    			tmp.clear();
    			lastlevel = level;
    		}
    		p = tq.front();
    		tq.pop();
    		tmp.push_back(p->val);
    		if (p->left != NULL)
    		{
    			tq.push(p->left);
    			lq.push(level + 1);
    		}

    		if (p->right != NULL)
    		{
    			tq.push(p->right);
    			lq.push(level + 1);
    		}
    	}	
    	if (res.size() < level)
    		res.push_back(tmp);
    	reverse(res.begin(), res.end());
    	return res;
    }
};

int main()
{

	return 0;
}

File: JavaSolution/src/BinaryTreeLevelOrderTraversal2.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Collections;

/**
 * Created by xyunpeng on 7/7/15.
 */
public class BinaryTreeLevelOrderTraversal2 {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        if (root == null)
            return res;
        q.add(root);
        List<Integer> el = new ArrayList<Integer>();
        int cur = 0;
        int last = 1;
        while(!q.isEmpty()) {
            TreeNode t = q.poll();
            last--;
            el.add(t.val);
            if (t.left != null) {
                cur++;
                q.add(t.left);
            }
            if (t.right != null) {
                cur++;
                q.add(t.right);
            }
            if (last == 0) {
                last = cur;
                cur = 0;
                res.add(el);
                el = new ArrayList<Integer>();
            }
        }
        Collections.reverse(res);
        return res;
    }
}
